### ISSUE TITLE
Memory leak: Publishing messages continuously leads to heap out of memory

### ISSUE THREAD

**Issue by ItsEcholot** [apollographql/graphql-subscriptions#97]
While benchmarking the graphql-subscriptions and subscriptions-transport-ws package I stumbled upon a potential memory leak.
The memory leak occurs if a client (in this case graphiql) is constantly connected and subscribed to the test topic while the server publishes random data to said topic.

**Intended outcome**
Graphql should continue to publish data to the topic without using more memory with each publish.
The test was done with this code:

_index.ts_
```Typescript
import * as bodyParser from 'body-parser';
import * as express from 'express';
import {execute, subscribe} from 'graphql';
import {graphiqlExpress, graphqlExpress} from 'graphql-server-express';
import {createServer} from 'http';
import {SubscriptionServer} from 'subscriptions-transport-ws';

const PORT = 3000;

const app = express();

app.use('/graphql', bodyParser.json(), graphqlExpress({
    schema: graphqlSchema,
}));
app.use('/graphiql', graphiqlExpress({
    endpointURL: '/graphql',
    subscriptionsEndpoint: `ws://localhost:${PORT}/graphql/subscriptions`,
}));

const ws = createServer(app);
ws.listen(PORT, () => {
    console.log('Listening on ' + PORT);

    // Setup WebSocket for handling GraphQL subscriptions
    const subServer = new SubscriptionServer({
        execute,
        schema: graphqlSchema,
        subscribe,
    }, {
        path: '/graphql/subscriptions',
        server: ws,
    });
});
```

_schema.ts_
```Typescript
import {makeExecutableSchema} from 'graphql-tools';
import resolvers from './resolvers';
const typeDefs = `
type Location {
    locationPLZ: Int!
    locationName: String!
    locationAddress: String
    locationPhone: String
    locationEMail: String
    locationOpeningHours: String
}

type ConfigurationItem {
    configurationItemId: Int!
    configurationItemName: String!
    configurationItemDescription: String
}

type DeviceModel {
    deviceModelId: Int!
    deviceModelName: String!
    
    deviceType: DeviceType!
}

type DeviceType {
    deviceTypeId: Int!
    deviceTypeName: String!
    deviceTypeIconName: String!
    templateName: String!
    startNumber: Int!
}

# Physical devices
type Device {
    # Serial number that identifies device
    deviceSerialNumber: String!
    deviceMacAddress: String
    deviceLocationDescription: String
    deviceDescription: String
    deviceName: String!
    
    location: Location
    deviceType: DeviceType
    deviceModel: DeviceModel
    configurationItem: ConfigurationItem
    logEntries: [LogEntry]
}

type LogCategory {
    logCategoryID: Int!
    logCategoryName: String!
    logCategoryLabel: String!
    
    logEntries: [LogEntry]!
}

type LogEntry {
    logEntryId: Int!
    logEditor: String!
    logEntry: String!
    logTimeStamp: String!
    
    device: Device!
    logCategory: LogCategory!
}



type Query {
    locationByPLZ(locationPLZ: Int): Location
    configurationItemById(configurationItemId: Int): ConfigurationItem
    deviceModelById(deviceModelId: Int): DeviceModel
    deviceTypeById(deviceTypeId: Int): DeviceType
    
    deviceBySerialNumber(deviceSerialNumber: String): Device
    devices: [Device]
}

type Subscription {
    deviceAdded: Device
}

schema {
  query: Query
  subscription: Subscription
}
`;

export default makeExecutableSchema({ typeDefs, resolvers });
```

_resolvers.ts_
```Typescript
import {PubSub} from 'graphql-subscriptions';

const pubSub = new PubSub();

const resolvers = {
    // A lot of resolvers for queries and relations emitted to save space
    Subscription: {
        deviceAdded: {
            subscribe: (obj: any, args: any, context: any, info: any) => pubSub.asyncIterator('deviceAdded'),
        },
    },
};

// Here comes the benchmark part
setInterval(() => {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

    for (let i = 0; i < 50; i++)
        text += possible.charAt(Math.floor(Math.random() * possible.length));

    pubSub.publish('deviceAdded', {
        deviceAdded: {
            deviceName: text,
            deviceSerialNumber: text,
        },
    });
}, 100);

export default resolvers;
```

**Actual outcome**
The benchmark worked stellar at the beginning, even with faster intervals. But after leaving it run for a few hours (with 5ms intervals) the node instance crashed with a JavaScript heap out of memory message.
I reduced the interval to 100ms to make sure the browser wasn't to slow to handle the incoming messages and let the instance run over the weekend and monitored the memory usage using memwatch-ng.
As expected the instance crashed again and memwatch-ng reported a heap growth of approximately 64 mb/hr at 100ms benchmark interval.

Here a few lines from the memwatch-ng logging and the eventual crash:

```
{ growth: 117084600,
  reason: 'heap growth over 5 consecutive GCs (1h 44m 32s) - 64.09 mb/hr' }
{ num_full_gc: 104,
  num_inc_gc: 15370,
  heap_compactions: 104,
  usage_trend: 18.1,
  estimated_base: 429235026,
  current_base: 1249398392,
  min: 22096488,
  max: 1249398392 }
{ num_full_gc: 105,
  num_inc_gc: 16727,
  heap_compactions: 105,
  usage_trend: 1076.1,
  estimated_base: 92668052,
  current_base: 1358532584,
  min: 22096488,
  max: 1358532584 }
{ num_full_gc: 106,
  num_inc_gc: 16743,
  heap_compactions: 106,
  usage_trend: 1080.6,
  estimated_base: 219389614,
  current_base: 1359883680,
  min: 22096488,
  max: 1359883680 }
{ num_full_gc: 107,
  num_inc_gc: 16753,
  heap_compactions: 107,
  usage_trend: 1082.2,
  estimated_base: 333506936,
  current_base: 1360562840,
  min: 22096488,
  max: 1360562840 }
{ num_full_gc: 108,
  num_inc_gc: 17417,
  heap_compactions: 108,
  usage_trend: 974.1,
  estimated_base: 12049953,
  current_base: 1413904408,
  min: 22096488,
  max: 1413904408 }





{ growth: 56740320,
  reason: 'heap growth over 5 consecutive GCs (50m 45s) - 63.97 mb/hr' }
{ num_full_gc: 109,
  num_inc_gc: 17435,
  heap_compactions: 109,
  usage_trend: 983.9,
  estimated_base: 152372248,
  current_base: 1415272904,
  min: 22096488,
  max: 1415272904 }
{ num_full_gc: 110,
  num_inc_gc: 17453,
  heap_compactions: 110,
  usage_trend: 990.4,
  estimated_base: 278803271,
  current_base: 1416682480,
  min: 22096488,
  max: 1416682480 }

<--- Last few GCs --->

76556274 ms: Mark-sweep 1375.1 (1406.1) -> 1375.1 (1406.1) MB, 1193.3 / 0.1 ms [allocation failure] [GC in old space requested].
76557462 ms: Mark-sweep 1375.1 (1406.1) -> 1375.0 (1406.1) MB, 1187.7 / 0.1 ms [allocation failure] [GC in old space requested].
76558653 ms: Mark-sweep 1375.0 (1406.1) -> 1375.0 (1406.1) MB, 1191.2 / 0.1 ms [last resort gc].
76559845 ms: Mark-sweep 1375.0 (1406.1) -> 1375.0 (1406.1) MB, 1192.0 / 0.1 ms [last resort gc].


<--- JS stacktrace --->

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

I noticed with further testing, that if you reconnect the client (i.e reload the graphiql window and/or rerun the subscription query) the JavaScript GC seems to clean up the growth and the heap returns to it's starting size.

**How to reproduce the issue**
Just publish as many events as possible and watch the heap size. The faster the publishing the faster the heap growth.

**Comment by NeoPhi** [comment]
A quick look at an allocation timeline shows a nesting of promises. Still investigating but I think the issue maybe with an included library.

**Comment by NeoPhi** [comment]
I opened https://github.com/leebyron/iterall/issues/24 as that looks to be the source of the memory leak as best as I can tell.

**Comment by sgtram** [comment]
Hey man, could you please update package dependencies?
iterall: 1.1.2

It has memory leak fix.
Thx

**Comment by NeoPhi** [comment]
Fixed in release 0.5.2.
